### DISCUSSION TITLE
Assessment of Design Quality

### DISCUSSION THREAD

**Discussion by jody** [MetroCS/redistricting#183] [General]
_In collaboration with your team, evaluate the quality of this project's software design._

**Add your _individual assessment_ of this project's software design quality to this discussion.**

Include the name of your team in your post.

This activity is associated with the CS3250 assignment [Assessment of Design Quality](https://msudenver.instructure.com/courses/53244/assignments/699220). No reflection section is required.

**Comment by AragornIISOA** [comment]
Team: Agile Devs In The Making
I would say some of the technologies involved in the project could stand to potentially be updated/upgraded to be more modern, though this is not a necessity, just one of many possibilities.  From a software design standpoint, I think the code is well designed and comes together cohesively for the most part, though there is still more to be done.  From a design standpoint of the code, because the main program passes current checkstyle conventions, this makes it easier to read, easier to maintain, and easier for newcomers to make contributions.
-Kyle Eden

**Comment by sarahbarnes1994** [comment]
I think that overall the efficiency isn't completely there. There is some vagueness in terms of the descriptions in the code blocks like pattern in StarGenerator.java. I think if we were to have a flow chart of how this code actually flows it may make a little more sense. 
The idea behind is is pretty good and the conventions are mostly there but it could use some revamping to make it more up to date.

**Comment by ColinLouis95** [comment]
Team: AgileProgrammingTeam
My assessment of this project's software design quality is overall the code is pretty coherent despite the complexity of the code base. I agree that the existing code can be updated to more modern design due to some files dating back to 2019.

**Comment by DanielShermanSchool** [comment]
Team: Agile Programming Team

The software design of the project seems "bloated." There is excessive use of memory via use of nested array-lists, and a more efficient use of memory might benefit the project in the future if needed.

The restriction to a square region is highly restrictive to the user and allowing for more freedom in region selection would improve the product greatly.

Build tools and tech (coco, ant) used are out of date and can cause issues on modern hardware.

Code itself is well documented and easy to read, however.

**Comment by ghost** [comment]
Team Name: Agile Software Devs
Name: Brady R

The quality of the code in MetroCS/Redistricting is variable but overall I think that the code is well written. Each method is commented with a detailed description and the variables are all labeled. The classes are divided fairly effectively and it is easy to understand what is happening and the goal of the design. There can still be room for improvement regarding the efficiency of the code and its overall capabilities, but these are things that are accomplished over time. The build tools that are used to compile and build the code are outdated and updating them should be considered. Overall I think that the code is good and highly maintainable but is held back by the quality of the buildtools.

**Comment by LingT03** [comment]
Team: Agile Devs In The Making 
We were not able to analyze the design of all the files however for the one that we did (AllDistrictGen.java), we noticed that the project has the potential for improvements. As for the code itself (to the best of my understanding) I think that the code is adequate because it works as intended, however, I did notice that some segments of code did not have comments paired to assist new users with understandability. As for the comments that were available I found them to be very helpful and cleared up the confusion.
- Ling Thang

**Comment by esundbla** [comment]
The quality of this project is definitely a multi-level.  On the surface the code design demonstrates good conventions.  Good object-oriented design protocol.  It passes its unit testing and implements testing methods and coding.  There are a variety of tools available for testing and allow for comprehensive response.

**Comment by TheMaxta** [comment]
Team: Cohesive Coders

We had a hard time quickly determining what specific variables were intended for. For example, in the StartGenerator file, there is a confusing variable named "n" and the documentation only refers to it as a pattern. This probably makes sense to the developer, but as a person reading through it takes a long time to get to the bottom of the intention. It could save a lot of development time if every important variable had a proper explanation.

**Comment by noah-fullerton** [comment]
Cohesive Coders:
Noah Fullerton:
The software follows standard object-oriented programming practices, with well-separated classes that properly hide their internal methods and fields; so the overall structure and hierarchy is well organized. Some of the individual function names and field names need to be more descriptive; and some of the logic is difficult to understand. Some of the code is far too complex, with functions of 500+ NPath complexity; also there are many unused local variables.

**Comment by ajhinds94** [comment]
Team: Cohesive Coders
In general the software design seems to be pretty solid.  The project uses object oriented programming well with separate classes for different tasks and everything.  There was some vagueness when it came to class descriptors, and this did hinder some of our efforts when it came to writing tests for these classes.  In addition, from what I saw in the StarGenerator class there could've been more efficient methods for traversing than using recursion as that is very memory use heavy. Many of the tests used through ant could have better descriptions as well.

**Comment by ataha2** [comment]
Team: AgileProgrammingTeam
Some of the code is 2-3 years old, I think it would be better if it is updated regularly, but in general the software design is pretty good.
Author: Ali Taha

**Comment by raffayqamar** [comment]
Name: Muhammad Qamar
Agile Software Devs
I would say that from the design perspective, the code passed all the tests, and it is fairly easy to understand for someone who is new and the code is organized. One thing that I'll say it needs improvement on is that the build tools involved in this project need to be updated since those are quite old.

**Comment by Iso-latte** [comment]
Team: Agile Software Devs
Name: Joaquin R. Trujillo

I believe the quality of this code is very well. My team and I have found mostly positive attributes of this software design. We enjoyed the author and doc tags that helped understand the code. We also thought that all of the classes and methods were named well. However, we did think that some variables were too vague to understand such as in Location.java in the compareTo method which the argument object is just named o. This o variable is then used throughout the class. Overall, I believe that this is a pretty good design.

**Comment by Leshy4** [comment]
Team: Agile Software Devs
Name: Chris Welch

I believe the design quality of this project is could have been improved a little because it seems that the tests weren't written before the code and now some accurate tests will fail, but overall, the names of the classes and methods had descriptive names most of the time and the general design is chiefly proficient.
